pkg/maps/nat/stats/stats_test.go

```diff
@@ -0,0 +1,132 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package stats
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/maps/nat"
+	"github.com/cilium/cilium/pkg/promise"
+	"github.com/cilium/cilium/pkg/testutils"
+	"github.com/cilium/cilium/pkg/tuple"
+	"github.com/cilium/cilium/pkg/types"
+	"github.com/cilium/cilium/pkg/u8proto"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/hivetest"
+	"github.com/cilium/hive/job"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_topk(t *testing.T) {
+	top5 := newTopK[key4](5)
+	for i := byte(0); i < 10; i++ {
+		ip := types.IPv4{10, 0, 0, i}
+		k := key4{
+			DestAddr: ip,
+		}
+		top5.Push(k, int(i))
+	}
+	out := []int{}
+	top5.popForEach(func(key key4, count, ith int) { out = append(out, count) })
+	assert.Equal(t, []int{5, 6, 7, 8, 9}, out)
+}
+
+func Test_countNat(t *testing.T) {
+	testutils.PrivilegedTest(t)
+
+	ip4Map := nat.NewMap("test_nat_map_ip4", nat.IPv4, 262144)
+	err := ip4Map.OpenOrCreate()
+	assert.NoError(t, err)
+	ip6Map := nat.NewMap("test_nat_map_ip6", nat.IPv6, 262144)
+	err = ip6Map.OpenOrCreate()
+	assert.NoError(t, err)
+	t.Cleanup(func() {
+		ip4Map.UnpinIfExists()
+		ip6Map.UnpinIfExists()
+	})
+
+	for addr := byte(0); addr < 20; addr++ {
+		for i := uint16(0); i < uint16(addr); i++ {
+			ip := types.IPv4{10, 0, 0, addr}
+			mapKey := &nat.NatKey4{}
+			mapKey.TupleKey4.SourceAddr = ip
+			mapKey.TupleKey4.DestAddr = [4]byte{}
+			mapKey.TupleKey4.DestPort = 9000 + i
+			mapKey.Flags = tuple.TUPLE_F_IN
+			mapKey.NextHeader = u8proto.TCP
+
```

Comment:
Could you spell `mapKey.TupleKey4.SourcePort = 0` out here (and below)? Or even use some other constant port number? It took me too long to grasp what's going on :grin:.

We could also add some identical UDP entries here (and/or vary the `SourcePort`). To emphasize that the statistics are truly sensitive to **any** difference in the `(SourceIP, SourcePort, DstIP, L4-Proto)` tuple, and we don't count false-positives.

Reply:
Good idea, I set them to mirror the endpoint IP somewhat by making them 8000 + <final_ip_octet>... ex:

```
10.0.0.1:8001 
10.0.0.1:8002
...
```